### PR TITLE
Allow ini-like filesize for RequestBodyBufferMiddleware and MultipartParser

### LINKED ISSUE #275: Allow ini-like filesizes
A few constructors allow to pass file sizes as parameters. In addition to bytes as integers, we should allow string values similar to the ini settings like `8M`.

Ref https://github.com/reactphp/http/pull/274#discussion_r155844986
Affects also https://github.com/reactphp/http/blob/master/src/Middleware/RequestBodyBufferMiddleware.php#L16

### PR DESCRIPTION
**Pull request by WyriHaximus**
Split off from #276 
Implements / closes #275

### CODE REVIEW

File: tests/Io/MultipartParserTest.php [hunk 1]
```diff
@@ -696,14 +696,14 @@ public function testUploadTooLargeFile()
         $data .= "Content-Disposition: form-data; name=\"file\"; filename=\"hello\"\r\n";
         $data .= "Content-type: text/plain\r\n";
         $data .= "\r\n";
-        $data .= "world\r\n";
+        $data .= str_repeat('world', 1024) . "\r\n";
```
Comment (clue) [inline tests/Io/MultipartParserTest.php]: The test makes perfect sense to me, but I feel that this is a bit misplaced here. Maybe add a separate test case for this? :+1:

File: src/Middleware/RequestBodyBufferMiddleware.php [hunk 2]
```diff
@@ -22,10 +22,10 @@
     public function __construct($sizeLimit = null)
     {
         if ($sizeLimit === null) {
-            $sizeLimit = IniUtil::iniSizeToBytes(ini_get('post_max_size'));
+            $sizeLimit = ini_get('post_max_size');
         }
 
-        $this->sizeLimit = $sizeLimit;
+        $this->sizeLimit = IniUtil::iniSizeToBytes($sizeLimit);
```
Comment (clue) [inline src/Middleware/RequestBodyBufferMiddleware.php]: LGTM, but doesn't match docblock/documentation.

File: src/Io/MultipartParser.php [hunk 3]
```diff
@@ -83,10 +83,10 @@ public function __construct($uploadMaxFilesize = null, $maxFileUploads = null)
         }
 
         if ($uploadMaxFilesize === null) {
-            $uploadMaxFilesize = IniUtil::iniSizeToBytes(ini_get('upload_max_filesize'));
+            $uploadMaxFilesize = ini_get('upload_max_filesize');
         }
 
-        $this->uploadMaxFilesize = (int)$uploadMaxFilesize;
+        $this->uploadMaxFilesize = IniUtil::iniSizeToBytes($uploadMaxFilesize);
```
Comment (clue) [inline src/Io/MultipartParser.php]: LGTM, but doesn't match docblock/documentation.

File: tests/Io/MultipartParserTest.php (hunk 1 above)
Comment (WyriHaximus) [reply tests/Io/MultipartParserTest.php]: Done :+1:

File: src/Middleware/RequestBodyBufferMiddleware.php (hunk 2 above)
Comment (WyriHaximus) [reply src/Middleware/RequestBodyBufferMiddleware.php]: Done :+1:

File: src/Io/MultipartParser.php (hunk 3 above)
Comment (WyriHaximus) [reply src/Io/MultipartParser.php]: Done :+1:

**Comment by WyriHaximus** [comment]
@clue updated and added docs and tests

**Comment by WyriHaximus** [comment]
And fixed the test error that just came up :blush:

File: tests/Middleware/RequestBodyBufferMiddlewareTest.php [hunk 4]
```diff
@@ -93,6 +93,33 @@ function (ServerRequestInterface $request) {
...
+
+        $stream = new ThroughStream();
+        $loop->addTimer(0.001, function () use ($stream) {
+            $stream->end(str_repeat('a', 2048));
+        });
+        $serverRequest = new ServerRequest(
+            'GET',
+            'https://example.com/',
+            array(),
+            new HttpBodyStream($stream, 2)
```
Comment (andig) [inline tests/Middleware/RequestBodyBufferMiddlewareTest.php]: Does the 2 make sense here?

File: src/Io/MultipartParser.php [hunk 5]
```diff
@@ -68,7 +68,7 @@
     private $emptyCount = 0;
 
     /**
-     * @param int|null $uploadMaxFilesize
+     * @param int|string|null $uploadMaxFilesize
```
Comment (clue) [inline src/Io/MultipartParser.php]: LGTM, same change is required for `RequestBodyParserMiddleware` :+1:

File: tests/Middleware/RequestBodyBufferMiddlewareTest.php (hunk 4 above)
Comment (clue) [reply tests/Middleware/RequestBodyBufferMiddlewareTest.php]: I agree, sementically no, doesn't make sense. Practically doesn't make a difference though.

File: src/Io/MultipartParser.php (hunk 5 above)
Comment (WyriHaximus) [reply src/Io/MultipartParser.php]: Done

File: tests/Middleware/RequestBodyBufferMiddlewareTest.php (hunk 4 above)
Comment (WyriHaximus) [reply tests/Middleware/RequestBodyBufferMiddlewareTest.php]: Fixed it anyway 👍

File: src/Middleware/RequestBodyBufferMiddleware.php [hunk 6]
```diff
@@ -14,18 +14,19 @@
     private $sizeLimit;
 
     /**
-     * @param int|null $sizeLimit Either an int with the max request body size
-     *                            or null to use post_max_size from PHP's
-     *                            configuration. (Note that the value from
-     *                            the CLI configuration will be used.)
+     * @param int|string|null $sizeLimit Either an int with the max request body size
+     *                                   or an ini like sze string
```
Comment (jsor) [inline src/Middleware/RequestBodyBufferMiddleware.php]: sze -> size. Also, maybe mention an example like `8M`.

File: src/Middleware/RequestBodyBufferMiddleware.php [hunk 7]
```diff
@@ -14,18 +14,19 @@
     private $sizeLimit;
 
     /**
-     * @param int|null $sizeLimit Either an int with the max request body size
-     *                            or null to use post_max_size from PHP's
-     *                            configuration. (Note that the value from
-     *                            the CLI configuration will be used.)
+     * @param int|string|null $sizeLimit Either an int with the max request body size
```
Comment (jsor) [inline src/Middleware/RequestBodyBufferMiddleware.php]: Maybe "...body size **in bytes**..."

File: src/Middleware/RequestBodyBufferMiddleware.php (hunk 6 above)
Comment (WyriHaximus) [reply src/Middleware/RequestBodyBufferMiddleware.php]: Fixed :+1:

File: src/Middleware/RequestBodyBufferMiddleware.php (hunk 7 above)
Comment (WyriHaximus) [reply src/Middleware/RequestBodyBufferMiddleware.php]: Good catch, done :+1:

**Pull request merged**